libvga/vga_driver.c

```diff
@@ -0,0 +1,149 @@
+#include "../common.h"
+#include "libvga.h"
+#include "../string/string.h"
+
+#define VGA_WIDTH 80
+#define VGA_HEIGHT 25
+
+u16 *vga_mem = (u16*)0xB8000;
+u32 col = 2; // Which column we're writing in
+u32 row = 0; // Which row we're writing in
+
+void vga_print_new_line() {
+    ++row;
+    col = 2;
+    if (row >= VGA_HEIGHT) {
+        for (int i = 1; i < VGA_HEIGHT; ++i) {
+            for (int j = 0; j < VGA_WIDTH; ++j) {
+                vga_mem[j + (i-1 * VGA_WIDTH)] = vga_mem[j + (i * VGA_WIDTH)];
+            }
+        }
+        for (int i = 0; i < VGA_WIDTH; ++i) {
+            vga_mem[(VGA_HEIGHT-1) + (i-1 * VGA_WIDTH)] = (Black << 8) | ' ';
+        }
+
+        vga_draw_boot_screen();
+    }
+}
+
+void vga_clear_screen() {
```

Comment:
This might be better done as a single loop. Because it's a u16, width * height * (implicit 2 byte stride) should bring you across the whole array, stopping at the end, and it needs a little less programmer read-time. Maybe something like this:
```
for (int i = 0; i < VGA_HEIGHT * VGA_WIDTH; i++) {
    vga_mem[i] = (Black << 8) | ' ';
}
```